core/src/main/java/hivemall/GeneralLearnerBaseUDTF.java

```diff
@@ -318,66 +412,240 @@ protected void onlineUpdate(@Nonnull final FeatureValue[] features, final float
-                final FloatWritable cov = new FloatWritable();
-                final IMapIterator<Object, IWeightValue> itor = model.entries();
-                while (itor.next() != -1) {
-                    itor.getValue(probe);
-                    if (!probe.isTouched()) {
-                        continue; // skip outputting untouched weights
+        finalizeTraining();
+        forwardModel();
+        this.accumulated = null;
+        this.model = null;
+    }
+
+    @VisibleForTesting
+    public void finalizeTraining() throws HiveException {
+        if (count == 0L) {
+            this.model = null;
+            return;
+        }
+        if (is_mini_batch) { // Update model with accumulated delta
+            batchUpdate();
+        }
+        if (iterations > 1) {
+            runIterativeTraining(iterations);
+        }
+    }
+
+    protected final void runIterativeTraining(@Nonnegative final int iterations)
+            throws HiveException {
+        final ByteBuffer buf = this.inputBuf;
+        final NioStatefullSegment dst = this.fileIO;
+        assert (buf != null);
+        assert (dst != null);
+        final long numTrainingExamples = count;
+
+        final Reporter reporter = getReporter();
+        final Counters.Counter iterCounter = (reporter == null) ? null : reporter.getCounter(
+            "hivemall.GeneralLearnerBase$Counter", "iteration");
+
+        try {
+            if (dst.getPosition() == 0L) {// run iterations w/o temporary file
+                if (buf.position() == 0) {
+                    return; // no training example
+                }
+                buf.flip();
+
+                int iter = 2;
+                double cumLossPrev;
+                for (; iter <= iterations; iter++) {
+                    cumLossPrev = cumLoss;
+                    this.cumLoss = 0.d;
+
+                    reportProgress(reporter);
+                    setCounterValue(iterCounter, iter);
+
+                    while (buf.remaining() > 0) {
+                        int recordBytes = buf.getInt();
+                        assert (recordBytes > 0) : recordBytes;
+                        int featureVectorLength = buf.getInt();
+                        final FeatureValue[] featureVector = new FeatureValue[featureVectorLength];
+                        for (int j = 0; j < featureVectorLength; j++) {
+                            featureVector[j] = new FeatureValue(buf);
+                        }
+                        float target = buf.getFloat();
+                        train(featureVector, target);
+                    }
+                    buf.rewind();
+
+                    if (is_mini_batch) { // Update model with accumulated delta
+                        batchUpdate();
+                    }
+
+                    logger.info("[iter " + iter + "] cumulative loss: " + cumLoss);
+
+                    if (Math.abs(cumLossPrev - cumLoss) < tol) {
```

Comment:
`cumLoss` is always 0 due to ` this.cumLoss = 0.d;`. So, this conversion check is not working properly.